lua/harpoon/init.lua

```diff
@@ -152,22 +157,28 @@ function Harpoon.setup(self, partial_config)
     ---TODO: should we go through every seen list and update its config?
 
     if self.hooks_setup == false then
-        vim.api.nvim_create_autocmd({ "BufLeave", "VimLeavePre" }, {
-            group = HarpoonGroup,
-            pattern = "*",
-            callback = function(ev)
-                self:_for_each_list(function(list, config)
-                    local fn = config[ev.event]
-                    if fn ~= nil then
-                        fn(ev, list)
-                    end
-
-                    if ev.event == "VimLeavePre" then
-                        self:sync()
+        vim.api.nvim_create_autocmd(
+            { "BufLeave", "VimLeavePre", "DirChanged" },
+            {
+                group = HarpoonGroup,
+                pattern = "*",
+                callback = function(ev)
+                    if ev.event == "DirChanged" then
+                        self:reset()
```

Comment:
Suggestion for minor technical improvement:

As far as I can tell from reading the code, this change seems to _reload_ the entire state. I believe that the data model supports having lists multiple lists by working dir, so perhaps rather than completely reloading state, just load the state for the new dir _if it doesn't already exist in memory_, to optimize loading? Then the new file is only loaded _the first time_ you switch to a new directory.

Having said that, the harpoon file is _so small_ that reading it in when changing tab or window is not a real problem. I don't experience any delays when using this branch, i.e. changing tabs is instantaneous.

Reply:
I'm inclined to leave it reloading from the disk completely. Reading from a file is fast enough the user shouldn't ever notice it, so it seems like a fairly negligible performance cost when considering what it would take to properly store all directory states in memory.

One big issue I could see with caching directory states in memory is garbage collection. When do we remove a directory's state from memory? Keeping them indefinitely is asking for memory leaks (albeit small) for anyone who a long running neovim instance when working with all their projects. It's also not obvious to me what a good heuristic would be to clear them out from memory (only keep 10..?, keep for a time..?).

Ultimately, I don't see the potential performance gains being worth complicating the state management logic further.